---
title: "VSX2_Analysis"
author:
  name: "Cody Ramirez"
  email: cody.ramirez@stjude.org
  affiliation: St. Jude Children's Research Hospital, Memphis, TN
date: "8/16/2021"
output: html_document
---


```{r}
#Loadds in important libraries for downstream analysis
library(Seurat)
library(Matrix)
library(ggplot2)
library(future)
library(sctransform)
library(leiden)
library(harmony)
library(Rcpp)
library(rliger)
library(cowplot)
library(patchwork)
library(SeuratWrappers)
library(dittoSeq)
library(stringr)
library(knitr)

# For consistency in tSNE/UMAP.
set.seed(1234)
# Sets the future settings for multicore processing
plan(strategy = "multiprocess", workers = 50)#Changing the current plan to access parallelization
options(parallelly.fork.enable = TRUE, future.globals.maxSize = +Inf, future.seed = TRUE) #enables forking in Rstudio (potentially unstable)
plan()

# Pull standard cell cycle genes for cell cycle scoring
# Had to change genes to title case for mice genes (Human genes are annotated in all CAPS)
s.genes <- str_to_title(cc.genes.updated.2019$s.genes)
g2m.genes <- str_to_title(cc.genes.updated.2019$g2m.genes)

# Identifies all samples in directory based on folder names
samples <- list.dirs(paste(full_path, "/CellRangerCount", sep = ""), recursive = F)

# Reads QC spreadsheet (columns are 'Sample', 'lower', 'upper', and 'mito')
qc_cutoff <- read.csv(paste(full_path, "/scRNAseq_QC_cutoff.tsv", sep = ""), sep = '\t', header = TRUE)


sample.name <- vector()
Seurat.obj <- list()


# For loops through all scRNA-Seq samples
for (i in 1:length(samples)){
  
  # Reads in CellRanger aligned data
  Seurat.data <- Read10X(data.dir = paste(full_path, "/CellRangerCount/", sample.name[i], "/outs/filtered_feature_bc_matrix", sep = ""))

  # Creates a Seurat Object, only includes genes expressed in at least 3 cells and includes cells where at least 200 genes were detected
  Seurat.obj[[i]] <- CreateSeuratObject(counts = Seurat.data, min.cells = 3, min.features = 200)

  # Adds sample name to metadata
  Seurat.obj[[i]]$Sample <- sample.name[i]

  # Calculate the percentage of reads that map to the mitochondrial genome
  Seurat.obj[[i]] <- PercentageFeatureSet(Seurat.obj[[i]], pattern = "^mt-", col.name = "percent.mito")
  
  # Visualize QC metric data before filtering is conducted
  VlnPlot(object = Seurat.obj[[i]], features = c("nCount_RNA", "nFeature_RNA", "percent.mito"), pt.size = 0, ncol = 3)
  FeatureScatter(object = Seurat.obj[[i]], feature1 = "nCount_RNA", feature2 = "percent.mito") + ggplot2::geom_hline(yintercept = qc_cutoff$mito[i])
  FeatureScatter(object = Seurat.obj[[i]], feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + ggplot2::geom_hline(yintercept = qc_cutoff$lower[i]) + ggplot2::geom_hline(yintercept = qc_cutoff$upper[i])

  # Filtering cells by # of genes and mitochondrial content, specific filtering parameters can be found in scRNAseq_QC_cutoff.tsv
  ############################################################################################################
  Seurat.obj[[i]] <- subset(x = Seurat.obj[[i]], subset = nFeature_RNA < qc_cutoff$upper[i] & nFeature_RNA > qc_cutoff$lower[i] & percent.mito < qc_cutoff$mito[i])
  ############################################################################################################
  
  # Visualize QC metric data after filtering is conducted
  VlnPlot(object = Seurat.obj[[i]], features = c("nCount_RNA", "nFeature_RNA", "percent.mito"), pt.size = 0, ncol = 3)
  FeatureScatter(object = Seurat.obj[[i]], feature1 = "nCount_RNA", feature2 = "percent.mito") + ggplot2::geom_hline(yintercept = qc_cutoff$mito[i])
  FeatureScatter(object = Seurat.obj[[i]], feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + ggplot2::geom_hline(yintercept = qc_cutoff$lower[i]) + ggplot2::geom_hline(yintercept = qc_cutoff$upper[i])

}
rm(qc_cutoff, Seurat.data)
```





```{r}
Single.Sample <- function(Seurat.obj)
{
  Seurat.obj <- NormalizeData(Seurat.obj)
  Seurat.obj <- CellCycleScoring(Seurat.obj, s.features = s.genes, g2m.features = g2m.genes)
  Seurat.obj <- FindVariableFeatures(Seurat.obj, selection.method = "vst", nfeatures = 3000)
  Seurat.obj <- ScaleData(Seurat.obj)
  Seurat.obj <- RunPCA(Seurat.obj, verbose = FALSE)
  Seurat.obj <- RunUMAP(Seurat.obj, dims = 1:30)
  Seurat.obj <- FindNeighbors(Seurat.obj, dims = 1:30)
  Seurat.obj <- FindClusters(Seurat.obj, resolution = 0.4, algorithm = 4)
  
  return(Seurat.obj)
}


Simple.Merge <- function(Seurat.list)
{

  Seurat.list <- lapply(X = Seurat.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- CellCycleScoring(x, s.features = s.genes, g2m.features = g2m.genes)
  })
  
  Seurat.combined <- merge(Seurat.list[[1]], Seurat.list[c(-1)])
  Seurat.combined <- FindVariableFeatures(Seurat.combined, selection.method = "vst", nfeatures = 3000)
  Seurat.combined <- ScaleData(Seurat.combined)
  Seurat.combined <- RunPCA(Seurat.combined, verbose = FALSE)
  Seurat.combined <- RunUMAP(Seurat.combined, dims = 1:30)
  Seurat.combined <- FindNeighbors(Seurat.combined, dims = 1:30)
  Seurat.combined <- FindClusters(Seurat.combined, resolution = 0.4, algorithm = 4)

  return(Seurat.combined)
}


Seurat.Integration <- function(Seurat.list)
{

  Seurat.list <- lapply(X = Seurat.list, FUN = function(x) {
  x <- SCTransform(x, variable.features.n = 3000, verbose = TRUE, return.only.var.genes = FALSE)
  x <- CellCycleScoring(x, s.features = s.genes, g2m.features = g2m.genes)
  })
  
  features <- SelectIntegrationFeatures(object.list = Seurat.list, nfeatures = 3000)
  Seurat.list <- PrepSCTIntegration(object.list = Seurat.list, anchor.features = features)
  anchors <- FindIntegrationAnchors(object.list = Seurat.list, normalization.method = "SCT", anchor.features = features)
  Seurat.combined <- IntegrateData(anchorset = anchors, normalization.method = "SCT")
  Seurat.combined <- RunPCA(Seurat.combined, verbose = FALSE)
  Seurat.combined <- RunUMAP(Seurat.combined, reduction = "pca", dims = 1:30)
  Seurat.combined <- FindNeighbors(Seurat.combined, dims = 1:30)
  Seurat.combined <- FindClusters(Seurat.combined, resolution = 0.4, algorithm = 4)

  return(Seurat.combined)
}


Harmony.Integration <- function(Seurat.list)
{

  Seurat.list <- lapply(X = Seurat.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- CellCycleScoring(x, s.features = s.genes, g2m.features = g2m.genes)
  })
  
  Seurat.combined <- merge(Seurat.list[[1]], Seurat.list[c(-1)])
  Seurat.combined <- FindVariableFeatures(Seurat.combined, selection.method = "vst", nfeatures = 3000)
  Seurat.combined <- ScaleData(Seurat.combined)
  Seurat.combined <- RunPCA(Seurat.combined, verbose = FALSE)
  Seurat.combined <- RunHarmony(Seurat.combined, group.by.vars = "Sample")
  Seurat.combined <- RunUMAP(Seurat.combined, reduction = "harmony", dims = 1:30)
  Seurat.combined <- FindNeighbors(Seurat.combined, reduction = "harmony", dims = 1:30)
  Seurat.combined <- FindClusters(Seurat.combined, resolution = 0.4, algorithm = 4)

  return(Seurat.combined)
}


Liger.Integration <- function(Seurat.list)
{

  Seurat.list <- lapply(X = Seurat.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- CellCycleScoring(x, s.features = s.genes, g2m.features = g2m.genes)
  })
  
  Seurat.combined <- merge(Seurat.list[[1]], Seurat.list[c(-1)])
  Seurat.combined <- FindVariableFeatures(Seurat.combined, selection.method = "vst", nfeatures = 3000)
  Seurat.combined <- ScaleData(Seurat.combined, split.by = "Sample", do.center = FALSE)
  Seurat.combined <- RunOptimizeALS(Seurat.combined, k = 20, lambda = 5, split.by = "Sample")
  Seurat.combined <- RunQuantileNorm(Seurat.combined, split.by = "Sample")
  # You can optionally perform Louvain clustering (`FindNeighbors` and `FindClusters`) after
  # `RunQuantileNorm` according to your needs
  Seurat.combined <- FindNeighbors(Seurat.combined, reduction = "iNMF", dims = 1:20)
  # Dimensional reduction and plotting
  Seurat.combined <- RunUMAP(Seurat.combined, dims = 1:ncol(Seurat.combined[["iNMF"]]), reduction = "iNMF")
  Seurat.combined <- FindClusters(Seurat.combined, resolution = 0.4, algorithm = 4)
  
  return(Seurat.combined)
}
```





```{r}
Generate.Integration.Plots <- function(Seurat.Exp, Experiment.Number, Integration.Method, Age)
{
  dir.create(paste(full_path, "Experiment#", Experiment.Number, sep = ""))
  dir.create(paste(full_path, "Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/", sep = ""))
  
  pdf(file = paste(full_path, "Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/1_", Integration.Method, "_integration_plots.pdf", sep = ""), width = 12, height = 6)
  p1 <- DimPlot(Seurat.Exp, reduction = "umap", group.by = "Sample")
  p2 <- DimPlot(Seurat.Exp, reduction = "umap", label = TRUE, repel = TRUE)
  print(p1 + p2)
  dev.off()
  
  total.width = length(unique(Seurat.Exp$Sample))*6
  pdf(file = paste(full_path, "Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/2_", Integration.Method, "_integration_plots.pdf", sep = ""), width = total.width, height = 6)
  print(DimPlot(Seurat.Exp, reduction = "umap", label = TRUE, split.by = "Sample"))
  dev.off()
}
```





```{r}
Generate.Feature.Plots <- function(Seurat.Exp, Experiment.Number, Integration.Method, Age)
{
  dir.create(paste(full_path, "/Experiment#", Experiment.Number, sep = ""))
  dir.create(paste(full_path, "/Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/", sep = ""))
  
  if (Age == 'Embryonic'){

    pdf(file = paste(full_path, "/Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/3_", Integration.Method, "_feature_plots.pdf", sep = ""), width = 12, height = 36)
    #Retinal progenitor cells (RPC):
    Retinal.Progenitor.Genes <- c("Pax6", "Cdk4", "Ccnd1", "Vsx2", "Lhx2", "Fgf15", "Sfrp2")
    Retinal.Progenitor.Genes <- Retinal.Progenitor.Genes[Retinal.Progenitor.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Retinal.Progenitor.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Neurogenic cells (characterized by expression of neurogenic bHLH factors):
    Neurogenic.Genes = c("Atoh7", "Olig2", "Neurog2", "Neurod1")
    Neurogenic.Genes <- Neurogenic.Genes[Neurogenic.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Neurogenic.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Retinal ganglion cells (RGC):
    Retinal.Ganglion.Genes = c("Thy1", "Nrn1", "Pou4f1", "Slc17a6", "Tusc5", "Rbpms")
    Retinal.Ganglion.Genes <- Retinal.Ganglion.Genes[Retinal.Ganglion.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Retinal.Ganglion.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Cone cells:
    Cone.Genes = c("Pde6c", "Pde6h", "Arr3", "Opn1sw", "Thrb", "Crx")
    Cone.Genes <- Cone.Genes[Cone.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Cone.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Amacrine cells:
    Amacrine.Genes = c("Tfap2a", "Tfap2b", "Pax6")
    Amacrine.Genes <- Amacrine.Genes[Amacrine.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Amacrine.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #RPE/endothelial cells:
    RPE.Endothelial.Genes = c("Mitf", "Msx1", "H19", "Cfh", "Pecam1", "Cspg4")
    RPE.Endothelial.Genes <- RPE.Endothelial.Genes[RPE.Endothelial.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, RPE.Endothelial.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Erythrocytes:
    Erythrocytes.Genes = c("Hba-a2", "Alas2", "Hbb-bt", "Fech")
    Erythrocytes.Genes <- Erythrocytes.Genes[Erythrocytes.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Erythrocytes.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    dev.off()
    
  }
  
  else if(Age == 'Adult'){
    
    pdf(file = paste(full_path, "/Experiment#", Experiment.Number, "/", Integration.Method, "_Integration/3_", Integration.Method, "_feature_plots.pdf", sep = ""), width = 12, height = 36)
    #Rod cells:
    Rod.Genes <- c('Nrl', 'Rho', 'Pdc', 'Pde6a', 'Pde6b')
    Rod.Genes <- Rod.Genes[Rod.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Rod.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Bipolar cells:
    Bipolar.Genes = c('Vsx2', 'Sebox', "Scgn")
    Bipolar.Genes <- Bipolar.Genes[Bipolar.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Bipolar.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Cone cells:
    Cone.Genes = c("Pde6c", "Pde6h", "Arr3")
    Cone.Genes <- Cone.Genes[Cone.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Cone.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))

    #Muller glia cells:
    Muller.Glia.Genes = c('Rlbp1', 'Rbp1', 'Cryab', 'Sox2')
    Muller.Glia.Genes <- Muller.Glia.Genes[Muller.Glia.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Muller.Glia.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #GABAergic amacrine cells:
    GABAergic.Amacrine.Genes = c('Gad1', 'Syt7', 'Slc6a1', 'Slc32a1')
    GABAergic.Amacrine.Genes <- GABAergic.Amacrine.Genes[GABAergic.Amacrine.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, GABAergic.Amacrine.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Glycinergic amacrine cells:
    Glycinergic.Amacrine.Genes = c('Slc6a9', 'Lamp5')
    Glycinergic.Amacrine.Genes <- Glycinergic.Amacrine.Genes[Glycinergic.Amacrine.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Glycinergic.Amacrine.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Retinal ganglion cells (RGC):
    Retinal.Ganglion.Genes = c("Thy1", "Nrn1", "Pou4f1", "Slc17a6", "Tusc5", "Rbpms")
    Retinal.Ganglion.Genes <- Retinal.Ganglion.Genes[Retinal.Ganglion.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Retinal.Ganglion.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Horizontal cells:
    Horizontal.Genes = c("Calb1", "Lhx1", "Oc3")
    Horizontal.Genes <- Horizontal.Genes[Horizontal.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Horizontal.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #RPE/endothelial cells:
    RPE.Endothelial.Genes = c("Mitf", "Msx1", "H19", "Cfh", "Pecam1", "Cspg4", "Cd34")
    RPE.Endothelial.Genes <- RPE.Endothelial.Genes[RPE.Endothelial.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, RPE.Endothelial.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))
    
    #Erythrocytes:
    Erythrocytes.Genes = c("Hba-a2", "Alas2", "Hbb-bt", "Fech")
    Erythrocytes.Genes <- Erythrocytes.Genes[Erythrocytes.Genes %in% getGenes(Seurat.Exp)]
    multi_dittoDimPlot(Seurat.Exp, Erythrocytes.Genes, split.by = 'Sample', ncol = 1, min.color = "gray90", max.color = "purple", theme = theme_classic() + theme(aspect.ratio = 1))

    dev.off()
  }
}
```
